### ISSUE TITLE
seqan2 fails to build on Linux/ia64

### ISSUE THREAD

**Issue by jduerstock** [seqan/seqan#2281]
Because ia64 uses "nop 0" rather than "nop", seqan fails to build on Linux/ia64.

From include/seqan/parallel/parallel_lock.h:
```
#else  // everything else.
    asm volatile ("nop" ::: "memory");  // default operation - does nothing => Might lead to passive spinning.
#endif
```

Is there a reason YieldProcessor() doesn't just use "{};" or "return;" rather than resorting to asm() calls when there's nothing to do?

**Comment by h-2** [comment]
We had actually had successful builds on ia64 before :thinking: 

@rrahn What's up with the asm nop, why do we need it?

**Comment by rrahn** [comment]
Hi, yes there is. We use a spin lock with hybrid spinning strategy (a combination of active and passive spinning). To prevent the compiler to optimise away the active spinning phase we use a instruction that does nothing but forces a tiny delay with a fixed number of idle cycles of the processor.
@jduerstock What you are referring to is actually a hard fallback, which might lead to unexpected longer waits than necessary for active spinning.
But this is only for architectures, we are not optimised for anyway. 
What I am wondering about is, why the `SSE2` macro is not detected (see below)? 
It should use the more performant `_mm_pause()` instruction here.
This should be available on IA64, shouldn't it? 
Can you please give us more detail about the architecture, the compiler and the compiler flags you used to compile SeqAn and which version of SeqAn you are referring too?
Did you use your own build infrastructure or our cmake build system?

```cpp
#elif defined(__SSE2__)  // AMD and Intel
    _mm_pause();
#else  // everything else.
    asm volatile ("nop" ::: "memory");  // default operation - does nothing => Might lead to passive spinning.
#endif
```

**Comment by jduerstock** [comment]
No, IA64 does not support SSE2.

This is building under Debian sid, which uses CMake and gcc 7.3.0.  The full build log is here:

https://buildd.debian.org/status/fetch.php?pkg=seqan2&arch=ia64&ver=2.4.0%2Bdfsg-5&stamp=1518354962&raw=0

**Comment by rrahn** [comment]
🤔 well, not sure why we did not see this before.
Can you make PR for IA64 with adding another macro check for IA64 before the SSE2 check adding the `nop 0` instruction.

**Comment by jduerstock** [comment]
IA64 was just recently reinstated as an unofficial architecture for debian-ports with in the past couple of months, which is probably why you hadn't heard about it before now.

I know you don't control the Debian packaging portion of things, but the current build status for seqan on all supported Debian architectures is available here: https://buildd.debian.org/status/package.php?p=seqan2&suite=sid

**Comment by mr-c** [comment]
Thank you @jduerstock for the context. I'm one of the Debian packagers for seqan and I've been working closely and extensively with @h-2, @rrahn, and the rest of the team on architecture support.

**Comment by mr-c** [comment]
> :thinking: well, not sure why we did not see this before.

Hey @rrahn , checking the build logs, `ia64` wasn't around for much of our testing and was failing the whole time when it was present
https://buildd.debian.org/status/logs.php?pkg=seqan2&arch=ia64
vs
https://buildd.debian.org/status/logs.php?pkg=seqan2&arch=amd64

**Comment by rrahn** [comment]
@mr-c many thanks for the info. I hope I'll can free some time next week and take a closer look.

**Comment by rrahn** [comment]
@mr-c that would be really great. This makes testing of the fix much more easier.
I look into it next week. many thanks.

**Comment by jrtc27** [comment]
> fixed number of idle cycles

Yeah that's not going to be true for any modern processor. Why not just do `asm volatile ("" ::: "memory");`?

**Comment by jduerstock** [comment]
This is still throwing errors six months later.  Can this be changed to `asm volatile ("" ::: "memory");` as suggested above by @jrtc27 ?

**Comment by h-2** [comment]
@rrahn @marehr Seems like a one line fix, can you please look into it?

**Comment by mr-c** [comment]
@h-2 @rrahn @marehr A fix that I've tested with qemu is available at https://github.com/seqan/seqan/pull/2370
